### PR TITLE
Github actions

### PR DESCRIPTION
**Pull request by ryanwinchester**
because why not

### CODE REVIEW

**Comment by dshoreman** [comment]
Thanks, will take a proper look over this later when I'm a bit more awake

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,20 +19,6 @@ jobs:
...
-          MYSQL_USER: root
-          MYSQL_PASSWORD: password
-          MYSQL_DATABASE: servidor_testing
-        ports: ['3306:3306']
-        options: >-
-          --health-cmd "mysqladmin ping"
-          --health-interval 5s
-          --health-timeout 2s
-          --health-retries 5
-
```
Comment (ryanwinchester) [inline .github/workflows/ci.yml]: You don't need a DB for tests?

File: .github/scripts/install.sh [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env bash
 
 # shellcheck source=retry.sh
-source "${CI_BUILD_DIR}/.github/scripts/retry.sh"
+source "${GITHUB_WORKSPACE}/.github/scripts/retry.sh"
```
Comment (ryanwinchester) [inline .github/scripts/install.sh:4]: P.S. I was trying to make everything generic enough that switching from gh actions to a different CI in the future, you'd only need to change the CI config

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dshoreman) [reply .github/workflows/ci.yml]: It's [supposedly preinstalled](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md#mysql) (MySQL rather than MariaDB, but 5.7 is probably good enough) so it's mostly to see if it works without.

Haven't checked the Windows env, but I'm not overly fussed if that doesn't work :wink:

File: .github/scripts/install.sh (hunk 2 above)
Comment (ryanwinchester) [reply .github/scripts/install.sh:4]: _which is why everything was_

```diff
- travis
+ CI
```

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ryanwinchester) [reply .github/workflows/ci.yml]: Ah, I didn't know this!

File: .github/scripts/install.sh (hunk 2 above)
Comment (dshoreman) [reply .github/scripts/install.sh:4]: Aye, that's fine - chances of moving again are incredibly slim so I'm not too worried. As long as GA stays free and working I see no reason to switch, and by moving ci-only stuff to .github/scripts it'll help clean things up overall.

A lot of the ugliness (eg `if [[ $RUN_MODE == "npm" ]]` and other no-ops) was largely due to Travis limitations. That or limitations in my knowledge of it. Essentially it proved difficult to separate npm and php steps entirely - leading to mess everywhere - which isn't an issue in GA it seems.

**Comment by dshoreman** [comment]
Well... I was rather hoping that a push would trigger the workflow since enabling Actions, but it's not detecting anything.

@ryanwinchester Are test runs not a thing before it's merged to the default branch?

**Comment by ryanwinchester** [comment]
> Well... I was rather hoping that a push would trigger the workflow since enabling Actions, but it's not detecting anything.
> 
> @ryanwinchester Are test runs not a thing before it's merged to the default branch?

You can just check here for now:

https://github.com/ryanwinchester/servidor/actions

**Comment by dshoreman** [comment]
Looks like we're almost there! With `--ignore-platform-req=php` on the composer install, PHP8 is failing as I'd expect, but that's fine. The main thing is we now know both JS+PHP parts of the workflow are fully functional.

**Some things left:**
* Composer and npm package caching
* Move scripts into the workflow
* Saw something about uploading log files, could be used for laravel log?
* Possibility of split stages (would require making `actions/checkout` persist somehow)

Anything else to consider? Currently Windows jobs claim to pass but actually do nothing, I suspect that'll change once the script commands shift into the workflow itself. The main benefit of merging the scripts though will be better control over how the Actions/workflow UI is rendered, separating each CS/SA tool into its own section etc.

**Comment by ryanwinchester** [comment]
You'll see I was playing around with matrices and cache keys. sorrynotsorry

I made some of the builds not count towards the overal success with a combination of `continue-on-error` and an `include` option in the matrix that overrides a variable that we can use as a flag.

@dshoreman are the cache directories correct? Should we also cache builds? (i.e. does webpack use incremental builds?)

**Comment by dshoreman** [comment]
> You'll see I was playing around with matrices and cache keys. sorrynotsorry

Haha, no worries. It'll get squashed anyway :)


> I made some of the builds not count towards the overal success with a combination of `continue-on-error` and an `include` option in the matrix that overrides a variable that we can use as a flag.

Thanks! Although 8.0 shouldn't allow failures (more on that at the end of the book) I *do* want to add `8.whatevers-next`, so I'll leave it all in place and we can just switch the versions out when the rest is done.


> @dshoreman are the cache directories correct? Should we also cache builds? (i.e. does webpack use incremental builds?)
> 
> Also we might want to make a list of cache `restore keys` with more lenient fallbacks

I'll double check the dirs shortly, but from what I remember it all looks good. WRT webpack, I have no clue - that's all a black box behind laravel mix, which I have no intention of messing with until my next attempt at a parcel switch :D 

Could you give a bit more background for listing the restore keys / fallback leniency? Haven't done much digging into the caching side of GHA yet, so I don't know much beyond "it's a thing"

---

###### the end of the book

I've added a couple commits earlier, one's just cleanup but the other moves everything from prepare.sh into the main install.sh script. I added a pair of `group_start "Some Group"` and `group_end` functions that *should* allow us to collapse sets of commands in each build step.

I also updated ramsey/uuid on the php8 branch just now, so that *should* be composer working. Obviously with my Travis credits maxed out we have no idea yet. I'll rebase this branch again, then we can probably drop `--ignore-platform-req=php` from the composer install too :tada:

**Comment by dshoreman** [comment]
Brief summary of changes since last time:

* Rebased again to (finally) fix random_compat. Forgot to `composer update` before :facepalm: 
* Had to dropp all Windows builds, sorry :(
  * They all either failed or simply didn't run any commands
  * Even if we spent a couple hours fixing that, tests would still fail due to Linux utils in the codebase
  * If people *really* want to develop on Windows, there's WSL
* Added Ubuntu 20.04
  * Currently failing because reasons unrelated to the action, will fix those post-merge
* Moved 8.0 back to top-level in matrix
  * PHPMD is currently failing on 8.0 - that's due to phpmd/phpmd#853, so we don't need to worry about it here
* Made PHP CS/SA and Testing two separate steps
  * Each tool is called (via make) directly in Workflow, no more script.sh!
  * phpmd reports "errors", but those are more "warnings" so it's fine. It'll only fail on "violations", IIRC.
* Updated actions/cache to v2
  * Added `restore_keys` for both node and composer
    * Falls back to `<cache_type>-<tool_version>-`
  * Removed os from created keys (without Windows, packages differences are unlikely)
* Added IRC notifications
  * These are *incredibly* basic atm, but it's "good enough" - see below

Think that covers most of it. Fixed a couple bugs in the Makefile caused by the OSX support so I'll update the changelog for that later.

Re notifications, GHA isn't *remotely* as flexible as Travis was, so I'll create an issue with the details to bring back full functionality of that separately. It's difficult to add a PR hook for it currently as it's only firing on push, which I expect will only change post-merge.

**Comment by ryanwinchester** [comment]
> Removed os from created keys (without Windows, packages differences are unlikely)

I don't know about PHP, but for Elixir things can behave very differently on Ubuntu 18.04 and 20.04. i.e. tests pass on one and fail on the other.

**Comment by dshoreman** [comment]
> > Removed os from created keys (without Windows, packages differences are unlikely)
> 
> I don't know about PHP, but for Elixir things can behave very differently on Ubuntu 18.04 and 20.04. i.e. tests pass on one and fail on the other.

tl;dr: Any local code directly calling `exec()` will still break tests, but none of the packages used are providing system functionality.  
We're also caching `~/.cache/composer` now instead of `vendor/`, so we always run `composer install` thus flagging any unresolvable dependencies. I can't foresee it causing issues but if it does, it'll be an easy fix when the time comes :smile: 

---

I guess the key difference for Elixir is it's far more likely to have calls to actual "system" functions which could be more prone to changing with the system. While the Servidor codebase itself is littered with `exec()` calls, that would be practically unheard of for a distributed Composer package.

The only ones I can really think of that vary deps between OS are file watchers but - even with those - all that matters is "is it Linux?" so they can require some inotify package for example. Pretty much everything else is abstracted in PHP functions, so 99% of packages have little reason to care if it's 16.04 or 24.04 as long as it has PHP X.Y

I could of course be talking from my rear, but until it becomes an issue I'm inclined to err on the side of "ehh, it'll be fiiiine!". Since it's just the cache key, worst case is it'll probably find it needs a different package and complain it's locked to a different version.

**Comment by dshoreman** [comment]
Is there anything else we're missing at this point? Aside from adding any extra steps (eg testing the installer) which can be done later, we gotta be getting pretty close now...

**Comment by ryanwinchester** [comment]
> I guess the key difference for Elixir is it's far more likely to have calls to actual "system" functions which could be more prone to changing with the system. While the Servidor codebase itself is littered with `exec()` calls, that would be practically unheard of for a distributed Composer package.

Most of the time it ends up being openssl issues between OS versions for me.

> Is there anything else we're missing at this point?

🚀

**Comment by dshoreman** [comment]
Thanks! It'd be nice to move the coverage report and Laravel log to final on-success/error steps, but uploading and downloading seems a bit of a faff from the docs. As it is, this does everything the travis config did and more, so I'm happy :)

Now to get back to fixing PHP 8 :joy:
